B3 Select instruction truncates vector operands

ddegazio opened the pull request:
#### c4e675b3afdd1aa16c3d3d3bd38c503bece0df0f
<pre>
B3 Select instruction truncates vector operands
<a href="https://bugs.webkit.org/show_bug.cgi?id=257842">https://bugs.webkit.org/show_bug.cgi?id=257842</a>
rdar://108643371

Reviewed by Mark Lam, Justin Michaud and Yusuke Suzuki.

Fixes a bug where we would generate a double-precision conditional
move when lowering the Select B3 instruction on vector operands.
Since vector-sized conditional move isn&apos;t widely supported, we
transform the Select into a branch and moves during Air lowering
when the operands are vectors.

* JSTests/wasm/stress/simd-select.js: Added.
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:

Canonical link: <a href="https://commits.webkit.org/264996@main">https://commits.webkit.org/264996@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c00b937df1d6ee2dfb1fede5d8d86b09ca7f8159

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9686 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9917 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11071 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11673 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9648 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12161 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9560 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/10479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11229 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/7774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/8584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8071 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/8871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/8733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12065 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9010 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9232 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/9586 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/8443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/2328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2276 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/12668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/9829 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8994 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2407 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

ddegazio commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -4469,6 +4469,25 @@ class LowerToAir {
+                ifTrueBlock->setSuccessors(doneBlock);
+                return;
+            }
+
```
For reviewers: this is the actual diff in `B3LowerToAir.cpp`, everything else is just trimmed trailing whitespace.

ddegazio replied:
Amended commit to remove all the whitespace changes.

hyjorc1 commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -4469,6 +4469,25 @@ class LowerToAir {
+                splitBlock(beginBlock, doneBlock);
+
+                append(Air::MoveVector, tmp(m_value->child(2)), tmp(m_value));
+                m_insts.last().append(createBranch(m_value->child(0)));
```
nit: Can we use `append()` instead of  `m_insts.last().append()`? Maybe we need a refactoring patch to replace all `m_insts.last().append()` with `append()`.

ddegazio replied:
Ah, I thought we couldn't do so, but it looks like we can. I'll change it!

ddegazio replied:
Changed in latest push.

MenloDorian commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -4469,6 +4469,25 @@ class LowerToAir {
         case Select: {
+            if (m_value->type().isVector()) {
+                // Conditional moves aren't widely supported for vectors, so
+                // we lower vector Select to a branching construct.
```
Since current architectures do not support conditional moves for vectors, can you change this comment to be more clear?  As is, it implies that some arch do support it but we're just going with the lowest common denominator here.  How about something like, "Conditional moves are available for vectors on currently supported architectures. So, we lower ..."?

MenloDorian replied:
Oops, I meant "Conditional moves are **not** available ..."

ddegazio replied:
Changed it in latest push.

MenloDorian reviewed (approved):
r=me

justinmichaud reviewed (approved):
r=me too

Constellation reviewed (approved):
r=me

ddegazio committed:
B3 Select instruction truncates vector operands
https://bugs.webkit.org/show_bug.cgi?id=257842
rdar://108643371

Reviewed by Mark Lam, Justin Michaud and Yusuke Suzuki.

Fixes a bug where we would generate a double-precision conditional
move when lowering the Select B3 instruction on vector operands.
Since vector-sized conditional move isn't widely supported, we
transform the Select into a branch and moves during Air lowering
when the operands are vectors.

* JSTests/wasm/stress/simd-select.js: Added.
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:

Canonical link: https://commits.webkit.org/264996@main
